### ISSUE TITLE
bug: Ionic overlays' focus return causes unexpected behavior

### ISSUE THREAD

**Issue by smishenk** [ionic-team/ionic-framework#22795]
# Bug Report

**Ionic version:**
[ ] **4.x**
[x] **5.3.4+**

**Current behavior:**
Our app has `IonActionSheet` with actions taking user to another page. During page navigation that action sheet appears for a moment in a broken state (let's say 'flashing').

**Expected behavior:**
No `IonActionSheet` 'flashing'

**Steps to reproduce:**
I don't think that you need to reproduce exactly my case here. Still I was able to find a root cause and it's connected with a fix you introduced in 5.3.4 version (https://github.com/ionic-team/ionic-framework/pull/22167). The page, user navigates to, displays `IonLoading` at the startup, which is dismissed (`onDidDismiss` of loading overlay is resolved) before the same thing from `IonActionSheet` overlay. So, at the moment `IonLoading` was _**presented**_, an `IonActionSheetButton` was an active element, therefore loading overlay returns focus to that button on dismiss and this results in displaying of `IonActionSheet` in some weird state.

**Other information:**
I wouldn't say it's a bug, but it's definitely an issue. Moreover, it might be more global than it seems.
Specifically in my case it's an issue with a pair of Ionic framework components. In the same time, if we take a look at such overlay behavior wider, you're playing with focus without any awareness of app layout and elements structure. Furthermore, you do this after some timeout (of dismiss animation), while other element may already have taken focus (intentionally) and you'll steal it again with no need at all. And speaking of code, I believe it is too much of responsibility for a common overlay component to give focus to whatever element it finds suitable (outside of overlay and modals/popovers/etc. wrapper components).
Well, it'd be nice to have at least a boolean flag to be able to disable such overlays behavior. In my view, it should be disabled by default, cause:

- It may not be needed at all (like in my case, i.e., user leaves a page - what's the point to give a focus back to invisible or deleted element?);
- It may be needed to give a focus to another element after overlay is dismissed (i.e., to set focus to the next form input after user filled the previous one);
- It causes issues with Ionic components itself;

That said, I'll be happy with any solution you come up to, which will provide control over overlays' focus return.

**Ionic info:**

```
Ionic:

   Ionic CLI                     : 6.12.3
   Ionic Framework               : @ionic/angular 5.3.4
   @angular-devkit/build-angular : 0.1100.7
   @angular-devkit/schematics    : 11.0.7
   @angular/cli                  : 11.0.7
   @ionic/angular-toolkit        : 3.0.0

Cordova:

   Cordova CLI       : 9.0.0 (cordova-lib@9.0.1)
   Cordova Platforms : none
   Cordova Plugins   : no whitelisted plugins (0 plugins total)

Utility:

   cordova-res : not installed
   native-run  : not installed

System:

   NodeJS : v12.18.1 (C:\Program Files\nodejs\node.exe)
   npm    : 6.14.5
   OS     : Windows 10
```

**Comment by liamdebeasi** [comment]
Thanks for the issue. This focus management is a common accessibility behavior designed to improve the usability of applications for folks who use assistive technology such as VoiceOver or TalkBack. That being said, there may be opportunities for improvement here. Can you reproduce this issue in an Ionic starter app and provide a link to the GitHub repo?

**Comment by smishenk** [comment]
> Thanks for the issue. This focus management is a common accessibility behavior designed to improve the usability of applications for folks who use assistive technology such as VoiceOver or TalkBack. That being said, there may be opportunities for improvement here. Can you reproduce this issue in an Ionic starter app and provide a link to the GitHub repo?

Thanks for your reply.
I got your point about assistive technology. I didn't mean that `focus return` is undesired in all of the cases. I tried to explain that managing `focus` is quite a complicated thing and I complained that such overlay behavior's been made default and there's no way to override or control it. There were a way to introduce the same `focus return` functionality, whatever app needs it, using overlay callbacks (to be honest, I'm pretty sure that it's the app itself must implement such functionality, not the framework).
Anyway, this issue is not only about this particular 'side effect' (that is just the first issue I ran into when updated Ionic). There ain't no doubts such overlay behavior introduces a lot of undesired and unexpected focus changes and keyboard show/hides, which is unacceptable for our app. And I suppose I'm not the only one. So, I don't get actually why you need a reproducer here? If you're going to fix exactly the bug I described, it'll be a waste of time for both of us))
I'm asking to put this overlay behavior under a boolean flag and this can be implemented without any reproducer, it's enough to convince you that it's highly required. Nevertheless, if you need a reproducer for some internal workflow requirements, I'll try to provide it using sample app.
Still, please take a second look at my points.

**Comment by liamdebeasi** [comment]
> I'm asking to put this overlay behavior under a boolean flag and this can be implemented without any reproducer, it's enough to convince you that it's highly required.

I do not think this is a good idea as this could result in all overlays no longer being accessible, which would be frustrating to those who use assistive technologies.

> So, I don't get actually why you need a reproducer here? If you're going to fix exactly the bug I described, it'll be a waste of time for both of us)

The reason why I asked for a code reproduction is that I cannot reproduce the issue as described. I am happy to make improvements to the accessibility focusing behavior, but I need to be able to reproduce the issue first.

**Comment by smishenk** [comment]
> I do not think this is a good idea as this could result in all overlays no longer being accessible, which would be frustrating to those who use assistive technologies.

Sorry, I don't understand how this can result in overlays being inaccessible. I'm talking about changes, like the following (_**core/src/utils/overlays.ts**_):
```javascript
if (opts.focusPreviousElementOnDismiss && overlay.el.tagName !== 'ION-TOAST') {
  focusPreviousElementOnDismiss(overlay.el);
}
```
That's it. It can be `true` by default (though I've already put my thoughts on it).
And let me focus on this part again: it was pretty easy to implement the same `focus return` functionality in any app that needed it even without your fix (thanks to nice overlay lifecycle hooks), but it's painful to override such forced `focus return` now in apps that don't need it in some cases or don't need it at all.

> The reason why I asked for a code reproduction is that I cannot reproduce the issue as described. I am happy to make improvements to the accessibility focusing behavior, but I need to be able to reproduce the issue first.

I see, I'll try to create some cases

**Comment by smishenk** [comment]
@liamdebeasi 
Here you are: https://github.com/smishenk/ionic-focus-issue
I was able to reproduce both cases in the latest Chrome version: 88.0.4324.96.
`Case1` goes for the issue described in this bug report.
`Case2` is aimed to illustrate another possible case, when focus is stolen from input on another page.
Both cases work correctly if you downgrade Ionic version to `5.3.3`.

**Comment by smishenk** [comment]
@liamdebeasi 
Did you get a chance to look at this issue?

**Comment by liamdebeasi** [comment]
Yes I did, but I am still investigating the issue.

**Comment by aeharding** [comment]
I would also like the option to disable return focus. I'd like to implement my own return focus logic for special situations.

My current situation is dismissing an action sheet does not correctly return focus to a textarea.

**Comment by liamdebeasi** [comment]
I spoke with the team, and we do not have plans to add this feature at the moment. The requested behavior is not found in the iOS/MD specs, so we think this is better implemented on a per-app basis.

However, we are interested in making it easier to pass custom HTML to overlays such as `ion-toast`. This would make it easy to pass an `ion-progress-bar` and control the element. You can follow https://github.com/ionic-team/ionic-framework/issues/18067 for updates on that. Thanks!
